File: AISpotlight/Providers/OpenAICompatibleProvider.swift
import Foundation

/// Chat provider for OpenAI-compatible APIs: OpenAI, Mistral, DeepSeek.
/// Uses `POST {base}/chat/completions` with SSE streaming, function calling,
/// and `GET {base}/models`.
struct OpenAICompatibleProvider: LLMProvider {
    let providerID: ProviderID
    let baseURL: URL

    static let openAI = OpenAICompatibleProvider(
        providerID: .openai,
        baseURL: URL(string: "https://api.openai.com/v1")!
    )
    static let mistral = OpenAICompatibleProvider(
        providerID: .mistral,
        baseURL: URL(string: "https://api.mistral.ai/v1")!
    )
    static let deepSeek = OpenAICompatibleProvider(
        providerID: .deepseek,
        baseURL: URL(string: "https://api.deepseek.com/v1")!
    )
    static let openRouter = OpenAICompatibleProvider(
        providerID: .openrouter,
        baseURL: URL(string: "https://openrouter.ai/api/v1")!
    )

    /// Builds a request with the Bearer key and any provider-specific headers.
    /// OpenRouter gets an `X-Title` attribution header for its app rankings.
    private func makeRequest(path: String, apiKey: String) -> URLRequest {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        if providerID == .openrouter {
            request.setValue("AISpotlight", forHTTPHeaderField: "X-Title")
        }
        return request
    }

    // MARK: - Chat

    func streamChat(
        messages: [LLMMessage],
        model: String,
        systemPrompt: String?,
        options: ChatRequestOptions,
        apiKey: String
    ) -> AsyncThrowingStream<LLMStreamEvent, Error> {
        // OpenAI's current primary endpoint is /v1/responses — unlike
        // /v1/chat/completions it supports function tools together with
        // reasoning. Mistral/DeepSeek stay on chat/completions.
        if providerID == .openai {
            return streamResponsesAPI(
                messages: messages,
                model: model,
                systemPrompt: systemPrompt,
                options: options,
                apiKey: apiKey
            )
        }
        var request = makeRequest(path: "chat/completions", apiKey: apiKey)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        var apiMessages: [[String: Any]] = []
        if let systemPrompt, !systemPrompt.isEmpty {
            apiMessages.append(["role": "system", "content": systemPrompt])
        }
        for message in messages {
            switch message.role {
            case .tool:
                apiMessages.append([
                    "role": "tool",
                    "tool_call_id": message.toolCallID ?? "",
                    "content": message.text
                ])
            case .assistant where !message.toolCalls.isEmpty:
                var entry: [String: Any] = [
                    "role": "assistant",
                    "tool_calls": message.toolCalls.map { call in
                        [
                            "id": call.id,
                            "type": "function",
                            "function": ["name": call.name, "arguments": call.argumentsJSON]
                        ]
                    }
                ]
                if !message.text.isEmpty { entry["content"] = message.text }
                apiMessages.append(entry)
            default:
                if message.images.isEmpty {
                    apiMessages.append(["role": message.role.rawValue, "content": message.text])
                } else {
                    var parts: [[String: Any]] = []
                    if !message.text.isEmpty {
                        parts.append(["type": "text", "text": message.text])
                    }
                    for image in message.images {
                        parts.append([
                            "type": "image_url",
                            "image_url": ["url": "data:\(image.mimeType);base64,\(image.base64)"]
                        ])
                    }
                    apiMessages.append(["role": message.role.rawValue, "content": parts])
                }
            }
        }

        var body: [String: Any] = [
            "model": model,
            "messages": apiMessages,
            "stream": true,
            "max_tokens": options.maxTokens
        ]
        if !options.tools.isEmpty {
            body["tools"] = options.tools.map { tool in
                [
                    "type": "function",
                    "function": [
                        "name": tool.name,
                        "description": tool.description,
                        "parameters": tool.parameters
                    ]
                ]
            }
        }
        // OpenRouter accepts a reasoning-effort knob directly on chat/completions
        // (Mistral/DeepSeek do not). Only sent for models the catalog says
        // support it, so it never reaches a model that would reject it.
        if providerID == .openrouter,
           options.reasoning != .auto,
           options.modelSupportsReasoning {
            body["reasoning"] = ["effort": options.reasoning == .fast ? "low" : "high"]
        }
        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        } catch {
            return AsyncThrowingStream { $0.finish(throwing: error) }
        }

        let sse = HTTPClient.sseStream(request)
        return AsyncThrowingStream { continuation in
            let task = Task {
                // Tool call deltas arrive fragmented — accumulate by index.
                var pendingCalls: [Int: (id: String, name: String, args: String)] = [:]
                do {
                    for try await payload in sse {
                        guard let data = payload.data(using: .utf8),
                              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { continue }

                        // OpenRouter (and some gateways) can surface a mid-stream
                        // error inside a data frame instead of a non-2xx status.
                        if let error = json["error"] as? [String: Any],
                           let message = error["message"] as? String {
                            throw ProviderError.http(status: (error["code"] as? Int) ?? 200, message: message)
                        }

                        guard let choices = json["choices"] as? [[String: Any]],
                              let choice = choices.first else { continue }

                        if let delta = choice["delta"] as? [String: Any] {
                            if let content = delta["content"] as? String, !content.isEmpty {
                                continuation.yield(.text(content))
                            }
                            if let toolCalls = delta["tool_calls"] as? [[String: Any]] {
                                for fragment in toolCalls {
                                    let index = fragment["index"] as? Int ?? 0
                                    var entry = pendingCalls[index] ?? (id: "", name: "", args: "")
                                    if let id = fragment["id"] as? String { entry.id = id }
                                    if let function = fragment["function"] as? [String: Any] {
                                        if let name = function["name"] as? String { entry.name += name }
                                        if let args = function["arguments"] as? String { entry.args += args }
                                    }
                                    pendingCalls[index] = entry
                                }
                            }
                        }
                    }
                    if !pendingCalls.isEmpty {
                        let calls = pendingCalls.sorted { $0.key < $1.key }.map { _, entry in
                            ToolCall(
                                id: entry.id.isEmpty ? UUID().uuidString : entry.id,
                                name: entry.name,
                                argumentsJSON: entry.args.isEmpty ? "{}" : entry.args
                            )
                        }
                        continuation.yield(.toolCalls(calls))
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    // MARK: - OpenAI Responses API

    /// `POST /v1/responses` with SSE streaming. Supports function tools
    /// together with reasoning (the chat/completions limitation doesn't apply).
    private func streamResponsesAPI(
        messages: [LLMMessage],
        model: String,
        systemPrompt: String?,
        options: ChatRequestOptions,
        apiKey: String
    ) -> AsyncThrowingStream<LLMStreamEvent, Error> {
        var request = URLRequest(url: baseURL.appendingPathComponent("responses"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")

        var input: [[String: Any]] = []
        for message in messages {
            switch message.role {
            case .tool:
                input.append([
                    "type": "function_call_output",
                    "call_id": message.toolCallID ?? "",
                    "output": message.text
                ])
            case .assistant where !message.toolCalls.isEmpty:
                if !message.text.isEmpty {
                    input.append([
                        "role": "assistant",
                        "content": [["type": "output_text", "text": message.text]]
                    ])
                }
                for call in message.toolCalls {
                    input.append([
                        "type": "function_call",
                        "call_id": call.id,
                        "name": call.name,
                        "arguments": call.argumentsJSON
                    ])
                }
            case .assistant:
                input.append([
                    "role": "assistant",
                    "content": [["type": "output_text", "text": message.text]]
                ])
            case .user:
                var parts: [[String: Any]] = []
                if !message.text.isEmpty {
                    parts.append(["type": "input_text", "text": message.text])
                }
                for image in message.images {
                    parts.append([
                        "type": "input_image",
                        "image_url": "data:\(image.mimeType);base64,\(image.base64)"
                    ])
                }
                input.append(["role": "user", "content": parts])
            }
        }

        var body: [String: Any] = [
            "model": model,
            "input": input,
            "stream": true,
            "store": false, // don't persist conversations server-side
            "max_output_tokens": options.maxTokens
        ]
        if let systemPrompt, !systemPrompt.isEmpty {
            body["instructions"] = systemPrompt
        }
        if !options.tools.isEmpty {
            // Responses API uses a flat function-tool shape.
            body["tools"] = options.tools.map { tool in
                [
                    "type": "function",
                    "name": tool.name,
                    "description": tool.description,
                    "parameters": tool.parameters
                ]
            }
        }
        if options.reasoning != .auto,
           ModelCapabilities.supportsReasoningControl(provider: .openai, model: model) {
            body["reasoning"] = ["effort": options.reasoning == .fast ? "low" : "high"]
        }

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        } catch {
            return AsyncThrowingStream { $0.finish(throwing: error) }
        }

        let sse = HTTPClient.sseStream(request)
        return AsyncThrowingStream { continuation in
            let task = Task {
                var pendingCalls: [ToolCall] = []
                do {
                    for try await payload in sse {
                        guard let data = payload.data(using: .utf8),
                              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                              let type = json["type"] as? String else { continue }

                        switch type {
                        case "response.output_text.delta":
                            if let delta = json["delta"] as? String, !delta.isEmpty {
                                continuation.yield(.text(delta))
                            }
                        case "response.output_item.done":
                            // Complete function call arrives assembled here.
                            if let item = json["item"] as? [String: Any],
                               item["type"] as? String == "function_call",
                               let name = item["name"] as? String {
                                pendingCalls.append(ToolCall(
                                    id: item["call_id"] as? String ?? UUID().uuidString,
                                    name: name,
                                    argumentsJSON: item["arguments"] as? String ?? "{}"
                                ))
                            }
                        case "response.failed", "response.incomplete":
                            if let response = json["response"] as? [String: Any],
                               let error = response["error"] as? [String: Any],
                               let message = error["message"] as? String {
                                throw ProviderError.http(status: 200, message: message)
                            }
                        case "error":
                            let message = json["message"] as? String ?? "Unknown streaming error"
                            throw ProviderError.http(status: 200, message: message)
                        default:
                            break
                        }
                    }
                    if !pendingCalls.isEmpty {
                        continuation.yield(.toolCalls(pendingCalls))
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    // MARK: - Models

    func fetchModels(apiKey: String) async throws -> [String] {
        let request = makeRequest(path: "models", apiKey: apiKey)

        let data = try await HTTPClient.json(request)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let items = json["data"] as? [[String: Any]] else {
            throw ProviderError.decoding("unexpected /models payload")
        }

        var ids: [String] = []
        for item in items {
            guard let id = item["id"] as? String else { continue }
            // Mistral advertises capabilities — keep chat-capable models only.
            if let caps = item["capabilities"] as? [String: Any],
               let chat = caps["completion_chat"] as? Bool, chat == false {
                continue
            }
            if isLikelyChatModel(id) {
                ids.append(id)
            }
        }
        return ids.sorted()
    }

    /// Filters out obviously non-chat models (embeddings, audio, images, moderation).
    private func isLikelyChatModel(_ id: String) -> Bool {
        let lower = id.lowercased()
        let excluded = [
            "embed", "whisper", "tts", "dall-e", "image", "moderation",
            "transcribe", "audio", "realtime", "ocr", "voxtral", "davinci",
            "babbage", "codestral-embed"
        ]
        return !excluded.contains { lower.contains($0) }
    }

    // MARK: - Model catalog (OpenRouter)

    /// Fetches the full model catalog with per-model capabilities, parsed from
    /// OpenRouter's richer `/models` payload (`architecture.input_modalities`,
    /// `supported_parameters`). Unfiltered on purpose: the catalog is used to
    /// validate whatever slug the user types, so it must accept every model
    /// OpenRouter serves. The key is optional (the endpoint is public).
    func fetchModelCatalog(apiKey: String?) async throws -> [ModelInfo] {
        var request = URLRequest(url: baseURL.appendingPathComponent("models"))
        if let apiKey { request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization") }
        if providerID == .openrouter {
            request.setValue("AISpotlight", forHTTPHeaderField: "X-Title")
        }

        let data = try await HTTPClient.json(request)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let items = json["data"] as? [[String: Any]] else {
            throw ProviderError.decoding("unexpected /models payload")
        }

        var catalog: [ModelInfo] = []
        for item in items {
            guard let id = item["id"] as? String else { continue }

            let inputModalities = (item["architecture"] as? [String: Any])?["input_modalities"] as? [String] ?? []
            let params = item["supported_parameters"] as? [String] ?? []
            catalog.append(ModelInfo(
                id: id,
                supportsVision: inputModalities.contains("image"),
                supportsTools: params.contains("tools"),
                supportsReasoning: params.contains("reasoning"),
                supportedParameters: params
            ))
        }
        return catalog
    }

    // MARK: - Key validation

    /// Verifies the API key with a cheap authenticated call. OpenRouter's
    /// `/models` is public, so a key check there would pass even with a bad key;
    /// `/api/v1/key` requires a valid key and returns its limits/usage.
    func validateKey(apiKey: String) async throws {
        if providerID == .openrouter {
            let request = makeRequest(path: "key", apiKey: apiKey)
            _ = try await HTTPClient.json(request)
        } else {
            _ = try await fetchModels(apiKey: apiKey)
        }
    }
}

File: android/app/src/main/kotlin/com/aispotlight/android/providers/OpenAICompatibleProvider.kt
package com.aispotlight.android.providers

import com.aispotlight.android.core.ChatRequestOptions
import com.aispotlight.android.core.HttpClient
import com.aispotlight.android.core.LLMMessage
import com.aispotlight.android.core.LLMProvider
import com.aispotlight.android.core.LLMStreamEvent
import com.aispotlight.android.core.ModelCapabilities
import com.aispotlight.android.core.ModelInfo
import com.aispotlight.android.core.ProviderException
import com.aispotlight.android.core.ProviderID
import com.aispotlight.android.core.ReasoningMode
import com.aispotlight.android.core.TokenUsage
import com.aispotlight.android.core.ToolCall
import com.aispotlight.android.core.newCallID
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONObject

/**
 * Chat provider for OpenAI-compatible APIs: OpenAI, Mistral, DeepSeek,
 * OpenRouter, Kimi (Moonshot).
 * Uses `POST {base}/chat/completions` with SSE streaming, function calling,
 * and `GET {base}/models`. OpenAI itself uses the newer `/v1/responses` API.
 */
class OpenAICompatibleProvider(
    override val providerID: ProviderID,
    private val baseURL: String,
) : LLMProvider {

    companion object {
        val openAI = OpenAICompatibleProvider(ProviderID.OPENAI, "https://api.openai.com/v1")
        val mistral = OpenAICompatibleProvider(ProviderID.MISTRAL, "https://api.mistral.ai/v1")
        val deepSeek = OpenAICompatibleProvider(ProviderID.DEEPSEEK, "https://api.deepseek.com/v1")
        val openRouter = OpenAICompatibleProvider(ProviderID.OPENROUTER, "https://openrouter.ai/api/v1")
        val kimi = OpenAICompatibleProvider(ProviderID.KIMI, "https://api.moonshot.ai/v1")
    }

    /**
     * Builds a request with the Bearer key and any provider-specific headers.
     * OpenRouter gets an `X-Title` attribution header for its app rankings.
     */
    private fun requestBuilder(path: String, apiKey: String): Request.Builder {
        val builder = Request.Builder()
            .url("$baseURL/$path")
            .header("Authorization", "Bearer $apiKey")
        if (providerID == ProviderID.OPENROUTER) {
            builder.header("X-Title", "AISpotlight")
        }
        return builder
    }

    // MARK: - Chat

    override fun streamChat(
        messages: List<LLMMessage>,
        model: String,
        systemPrompt: String?,
        options: ChatRequestOptions,
        apiKey: String,
    ): Flow<LLMStreamEvent> {
        // OpenAI's current primary endpoint is /v1/responses — unlike
        // /v1/chat/completions it supports function tools together with
        // reasoning. Mistral/DeepSeek stay on chat/completions.
        if (providerID == ProviderID.OPENAI) {
            return streamResponsesAPI(messages, model, systemPrompt, options, apiKey)
        }
        return flow {
            val apiMessages = JSONArray()
            if (!systemPrompt.isNullOrEmpty()) {
                apiMessages.put(JSONObject().put("role", "system").put("content", systemPrompt))
            }
            for (message in messages) {
                when {
                    message.role == LLMMessage.Role.TOOL -> {
                        apiMessages.put(JSONObject().apply {
                            put("role", "tool")
                            put("tool_call_id", message.toolCallID ?: "")
                            put("content", message.text)
                        })
                    }
                    message.role == LLMMessage.Role.ASSISTANT && message.toolCalls.isNotEmpty() -> {
                        val entry = JSONObject().apply {
                            put("role", "assistant")
                            put("tool_calls", JSONArray().apply {
                                for (call in message.toolCalls) {
                                    put(JSONObject().apply {
                                        put("id", call.id)
                                        put("type", "function")
                                        put("function", JSONObject()
                                            .put("name", call.name)
                                            .put("arguments", call.argumentsJSON))
                                    })
                                }
                            })
                        }
                        if (message.text.isNotEmpty()) entry.put("content", message.text)
                        apiMessages.put(entry)
                    }
                    else -> {
                        if (message.images.isEmpty()) {
                            apiMessages.put(JSONObject().put("role", message.role.raw).put("content", message.text))
                        } else {
                            val parts = JSONArray()
                            if (message.text.isNotEmpty()) {
                                parts.put(JSONObject().put("type", "text").put("text", message.text))
                            }
                            for (image in message.images) {
                                parts.put(JSONObject().apply {
                                    put("type", "image_url")
                                    put("image_url", JSONObject()
                                        .put("url", "data:${image.mimeType};base64,${image.base64}"))
                                })
                            }
                            apiMessages.put(JSONObject().put("role", message.role.raw).put("content", parts))
                        }
                    }
                }
            }

            val body = JSONObject().apply {
                put("model", model)
                put("messages", apiMessages)
                put("stream", true)
                put("max_tokens", options.maxTokens)
                if (options.tools.isNotEmpty()) {
                    put("tools", JSONArray().apply {
                        for (tool in options.tools) {
                            put(JSONObject().apply {
                                put("type", "function")
                                put("function", JSONObject().apply {
                                    put("name", tool.name)
                                    put("description", tool.description)
                                    put("parameters", tool.parameters)
                                })
                            })
                        }
                    })
                }
                // OpenRouter accepts a reasoning-effort knob directly on
                // chat/completions (Mistral/DeepSeek do not). Only sent for
                // models the catalog says support it.
                if (providerID == ProviderID.OPENROUTER &&
                    options.reasoning != ReasoningMode.AUTO &&
                    options.modelSupportsReasoning
                ) {
                    put("reasoning", JSONObject()
                        .put("effort", if (options.reasoning == ReasoningMode.FAST) "low" else "high"))
                }
                // Ask for the final usage chunk (cost tracking). Sent to
                // providers that document `stream_options`: DeepSeek, OpenRouter,
                // Kimi. Mistral is excluded until verified live — unknown
                // params risk a 422 there.
                if (providerID == ProviderID.DEEPSEEK ||
                    providerID == ProviderID.OPENROUTER ||
                    providerID == ProviderID.KIMI
                ) {
                    put("stream_options", JSONObject().put("include_usage", true))
                }
            }

            val request = requestBuilder("chat/completions", apiKey)
                .header("Content-Type", "application/json")
                .post(HttpClient.jsonBody(body))
                .build()

            // Tool call deltas arrive fragmented — accumulate by index.
            data class Pending(var id: String = "", var name: String = "", val args: StringBuilder = StringBuilder())
            val pendingCalls = sortedMapOf<Int, Pending>()
            var usage = TokenUsage()

            HttpClient.sseStream(request).collect { payload ->
                val json = try { JSONObject(payload) } catch (_: Exception) { return@collect }

                // OpenRouter (and some gateways) can surface a mid-stream error
                // inside a data frame instead of a non-2xx status.
                json.optJSONObject("error")?.let { error ->
                    val message = error.optString("message")
                    if (message.isNotEmpty()) {
                        throw ProviderException.http(error.optInt("code", 200), message)
                    }
                }

                // Usage rides on the final chunk, whose `choices` is empty —
                // must be read BEFORE the choice guard below skips it.
                json.optJSONObject("usage")?.let { usage = parseChatCompletionsUsage(it) }

                val choice = json.optJSONArray("choices")?.optJSONObject(0) ?: return@collect
                val delta = choice.optJSONObject("delta") ?: return@collect
                val content = delta.optString("content")
                if (content.isNotEmpty()) emit(LLMStreamEvent.Text(content))
                delta.optJSONArray("tool_calls")?.let { toolCalls ->
                    for (i in 0 until toolCalls.length()) {
                        val fragment = toolCalls.optJSONObject(i) ?: continue
                        val index = fragment.optInt("index", 0)
                        val entry = pendingCalls.getOrPut(index) { Pending() }
                        fragment.optString("id").takeIf { it.isNotEmpty() }?.let { entry.id = it }
                        fragment.optJSONObject("function")?.let { function ->
                            entry.name += function.optString("name")
                            entry.args.append(function.optString("arguments"))
                        }
                    }
                }
            }
            if (pendingCalls.isNotEmpty()) {
                emit(LLMStreamEvent.ToolCalls(pendingCalls.values.map { entry ->
                    ToolCall(
                        id = entry.id.ifEmpty { newCallID() },
                        name = entry.name,
                        argumentsJSON = entry.args.toString().ifEmpty { "{}" }
                    )
                }))
            }
            if (!usage.isEmpty) emit(LLMStreamEvent.Usage(usage))
        }
    }

    /**
     * Maps a chat/completions `usage` object to normalized counts.
     * DeepSeek splits input by cache outcome (hit ≈ 1/50 of miss price) —
     * its `prompt_cache_hit_tokens`/`prompt_cache_miss_tokens` take priority.
     * Other providers report `prompt_tokens` (total) + optional
     * `prompt_tokens_details.cached_tokens`; the cached share is subtracted so
     * `inputTokens` stays "uncached input" across all providers.
     */
    private fun parseChatCompletionsUsage(u: JSONObject): TokenUsage {
        val prompt = u.optInt("prompt_tokens")
        val completion = u.optInt("completion_tokens")
        val hit = u.optInt("prompt_cache_hit_tokens", -1)
        val miss = u.optInt("prompt_cache_miss_tokens", -1)
        val reasoning = u.optJSONObject("completion_tokens_details")?.optInt("reasoning_tokens") ?: 0
        return if (hit >= 0 && miss >= 0) {
            TokenUsage(inputTokens = miss, outputTokens = completion,
                cacheReadTokens = hit, reasoningTokens = reasoning)
        } else {
            val cached = u.optJSONObject("prompt_tokens_details")?.optInt("cached_tokens") ?: 0
            TokenUsage(inputTokens = (prompt - cached).coerceAtLeast(0), outputTokens = completion,
                cacheReadTokens = cached, reasoningTokens = reasoning)
        }
    }

    // MARK: - OpenAI Responses API

    /**
     * `POST /v1/responses` with SSE streaming. Supports function tools together
     * with reasoning (the chat/completions limitation doesn't apply).
     */
    private fun streamResponsesAPI(
        messages: List<LLMMessage>,
        model: String,
        systemPrompt: String?,
        options: ChatRequestOptions,
        apiKey: String,
    ): Flow<LLMStreamEvent> = flow {
        val input = JSONArray()
        for (message in messages) {
            when {
                message.role == LLMMessage.Role.TOOL -> {
                    input.put(JSONObject().apply {
                        put("type", "function_call_output")
                        put("call_id", message.toolCallID ?: "")
                        put("output", message.text)
                    })
                }
                message.role == LLMMessage.Role.ASSISTANT && message.toolCalls.isNotEmpty() -> {
                    if (message.text.isNotEmpty()) {
                        input.put(JSONObject().apply {
                            put("role", "assistant")
                            put("content", JSONArray().put(JSONObject()
                                .put("type", "output_text").put("text", message.text)))
                        })
                    }
                    for (call in message.toolCalls) {
                        input.put(JSONObject().apply {
                            put("type", "function_call")
                            put("call_id", call.id)
                            put("name", call.name)
                            put("arguments", call.argumentsJSON)
                        })
                    }
                }
                message.role == LLMMessage.Role.ASSISTANT -> {
                    input.put(JSONObject().apply {
                        put("role", "assistant")
                        put("content", JSONArray().put(JSONObject()
                            .put("type", "output_text").put("text", message.text)))
                    })
                }
                else -> {
                    val parts = JSONArray()
                    if (message.text.isNotEmpty()) {
                        parts.put(JSONObject().put("type", "input_text").put("text", message.text))
                    }
                    for (image in message.images) {
                        parts.put(JSONObject().apply {
                            put("type", "input_image")
                            put("image_url", "data:${image.mimeType};base64,${image.base64}")
                        })
                    }
                    input.put(JSONObject().put("role", "user").put("content", parts))
                }
            }
        }

        val body = JSONObject().apply {
            put("model", model)
            put("input", input)
            put("stream", true)
            put("store", false) // don't persist conversations server-side
            put("max_output_tokens", options.maxTokens)
            if (!systemPrompt.isNullOrEmpty()) put("instructions", systemPrompt)
            if (options.tools.isNotEmpty()) {
                // Responses API uses a flat function-tool shape.
                put("tools", JSONArray().apply {
                    for (tool in options.tools) {
                        put(JSONObject().apply {
                            put("type", "function")
                            put("name", tool.name)
                            put("description", tool.description)
                            put("parameters", tool.parameters)
                        })
                    }
                })
            }
            if (options.reasoning != ReasoningMode.AUTO &&
                ModelCapabilities.supportsReasoningControl(ProviderID.OPENAI, model)
            ) {
                put("reasoning", JSONObject()
                    .put("effort", if (options.reasoning == ReasoningMode.FAST) "low" else "high"))
            }
        }

        val request = Request.Builder()
            .url("$baseURL/responses")
            .header("Content-Type", "application/json")
            .header("Authorization", "Bearer $apiKey")
            .post(HttpClient.jsonBody(body))
            .build()

        val pendingCalls = mutableListOf<ToolCall>()
        var usage = TokenUsage()
        HttpClient.sseStream(request).collect { payload ->
            val json = try { JSONObject(payload) } catch (_: Exception) { return@collect }
            when (json.optString("type")) {
                "response.completed" -> {
                    // Whole-response usage arrives on the terminal event.
                    json.optJSONObject("response")?.optJSONObject("usage")?.let { u ->
                        val input = u.optInt("input_tokens")
                        val cached = u.optJSONObject("input_tokens_details")?.optInt("cached_tokens") ?: 0
                        usage = TokenUsage(
                            inputTokens = (input - cached).coerceAtLeast(0),
                            outputTokens = u.optInt("output_tokens"),
                            cacheReadTokens = cached,
                            reasoningTokens = u.optJSONObject("output_tokens_details")?.optInt("reasoning_tokens") ?: 0,
                        )
                    }
                }
                "response.output_text.delta" -> {
                    val delta = json.optString("delta")
                    if (delta.isNotEmpty()) emit(LLMStreamEvent.Text(delta))
                }
                "response.output_item.done" -> {
                    // Complete function call arrives assembled here.
                    val item = json.optJSONObject("item")
                    if (item?.optString("type") == "function_call") {
                        val name = item.optString("name")
                        if (name.isNotEmpty()) {
                            pendingCalls.add(ToolCall(
                                id = item.optString("call_id").ifEmpty { newCallID() },
                                name = name,
                                argumentsJSON = item.optString("arguments").ifEmpty { "{}" }
                            ))
                        }
                    }
                }
                "response.failed", "response.incomplete" -> {
                    val message = json.optJSONObject("response")
                        ?.optJSONObject("error")?.optString("message")
                    if (!message.isNullOrEmpty()) throw ProviderException.http(200, message)
                }
                "error" -> {
                    throw ProviderException.http(200, json.optString("message").ifEmpty { "Unknown streaming error" })
                }
            }
        }
        if (pendingCalls.isNotEmpty()) {
            emit(LLMStreamEvent.ToolCalls(pendingCalls))
        }
        if (!usage.isEmpty) emit(LLMStreamEvent.Usage(usage))
    }

    // MARK: - Models

    override suspend fun fetchModels(apiKey: String): List<String> {
        val request = requestBuilder("models", apiKey).build()
        val data = HttpClient.json(request)
        val items = try {
            JSONObject(data).getJSONArray("data")
        } catch (_: Exception) {
            throw ProviderException.decoding("unexpected /models payload")
        }

        val ids = mutableListOf<String>()
        for (i in 0 until items.length()) {
            val item = items.optJSONObject(i) ?: continue
            val id = item.optString("id").takeIf { it.isNotEmpty() } ?: continue
            // Mistral advertises capabilities — keep chat-capable models only.
            val caps = item.optJSONObject("capabilities")
            if (caps != null && caps.has("completion_chat") && !caps.optBoolean("completion_chat", true)) continue
            if (isLikelyChatModel(id)) ids.add(id)
        }
        return ids.sorted()
    }

    /** Filters out obviously non-chat models (embeddings, audio, images, moderation). */
    private fun isLikelyChatModel(id: String): Boolean {
        val lower = id.lowercase()
        val excluded = listOf(
            "embed", "whisper", "tts", "dall-e", "image", "moderation",
            "transcribe", "audio", "realtime", "ocr", "voxtral", "davinci",
            "babbage", "codestral-embed"
        )
        return excluded.none { lower.contains(it) }
    }

    // MARK: - Model catalog (OpenRouter)

    /**
     * Fetches the full model catalog with per-model capabilities, parsed from
     * OpenRouter's richer `/models` payload. Unfiltered on purpose: the catalog
     * validates whatever slug the user types. The key is optional (public endpoint).
     */
    suspend fun fetchModelCatalog(apiKey: String?): List<ModelInfo> {
        val builder = Request.Builder().url("$baseURL/models")
        if (apiKey != null) builder.header("Authorization", "Bearer $apiKey")
        if (providerID == ProviderID.OPENROUTER) builder.header("X-Title", "AISpotlight")

        val data = HttpClient.json(builder.build())
        val items = try {
            JSONObject(data).getJSONArray("data")
        } catch (_: Exception) {
            throw ProviderException.decoding("unexpected /models payload")
        }

        val catalog = mutableListOf<ModelInfo>()
        for (i in 0 until items.length()) {
            val item = items.optJSONObject(i) ?: continue
            val id = item.optString("id").takeIf { it.isNotEmpty() } ?: continue
            val inputModalities = item.optJSONObject("architecture")
                ?.optJSONArray("input_modalities")?.toStringList() ?: emptyList()
            val params = item.optJSONArray("supported_parameters")?.toStringList() ?: emptyList()
            // OpenRouter reports USD-per-token prices as decimal strings.
            val pricing = item.optJSONObject("pricing")
            catalog.add(ModelInfo(
                id = id,
                supportsVision = inputModalities.contains("image"),
                supportsTools = params.contains("tools"),
                supportsReasoning = params.contains("reasoning"),
                supportedParameters = params,
                promptPricePerToken = pricing?.optString("prompt")?.toDoubleOrNull(),
                completionPricePerToken = pricing?.optString("completion")?.toDoubleOrNull(),
            ))
        }
        return catalog
    }

    // MARK: - Key validation

    /**
     * OpenRouter's `/models` is public, so a key check there would pass even
     * with a bad key; `/api/v1/key` requires a valid key.
     */
    override suspend fun validateKey(apiKey: String) {
        if (providerID == ProviderID.OPENROUTER) {
            HttpClient.json(requestBuilder("key", apiKey).build())
        } else {
            fetchModels(apiKey)
        }
    }
}

private fun JSONArray.toStringList(): List<String> =
    (0 until length()).mapNotNull { optString(it).takeIf { s -> s.isNotEmpty() } }
